services/src/handlers/datasets.rs

```diff
@@ -51,12 +57,45 @@ where
         web::scope("/dataset")
             .service(web::resource("/suggest").route(web::get().to(suggest_meta_data_handler::<C>)))
             .service(web::resource("/auto").route(web::post().to(auto_create_dataset_handler::<C>)))
+            .service(web::resource("/volumes").route(web::get().to(list_volumes_handler)))
+            .service(
+                web::resource("/public").route(web::post().to(create_system_dataset_handler::<C>)),
```

Comment:
This sounds like a temporary solution to do it under `/public`.

Wouldn't it be better to treat user and public datasets the same and just give them other access rights?

Reply:
unfortunately it doesn't work to have the same route, and use a different handler depending on the kind of session token. I tried a single handler with Either<Admin, User> and Either<SystemData, UserData> but it wouldn't match. I also tried two services one with admin and one with regular session, but it would always match the first one and throw an error if the session is not an admin one e.g.

Do you maybe have an idea of how to do this?

Reply:
I guess `Either` is only about being able to deserialize stuff.
So, are both things differentiable by their inner schema?

Reply:
what is an inner schema? The CreateUserDataset and CreateSystemDataset have differently named fields. Session and AdminSession are identical though.

Reply:
But `AdminSession`  has a custom deserializer.

So, is there an error why this does not match?

Reply:
I tried

```
pub async fn create_user_dataset_handler<C: Context>(
    session: Either<C::Session, AdminSession>,
    ctx: web::Data<C>,
    create: Either<web::Json<CreateUserDataset>, web::Json<CreateSystemDataset>>,
) -> Result<web::Json<IdResponse<DatasetId>>> {
    let (session, create) = match (session, create) {
        (Either::Left(session), Either::Left(create)) => ...
        (Either::Right(session), Either::Right(create)) => ...
```` 

but then I get

> [services/src/handlers/datasets.rs:1143] res.response().body() = BoxBody(
    Bytes(
        b"{\"error\":\"BodyDeserializeError\",\"message\":\"EOF while parsing a value at line 1 column 0\"}",
    ),
)

so actix can't seem to handle the `Either`in the body? I don't know.

Reply:
Does Either work for web::Json at all?

Reply:
Yeah, that also sounds reasonable. I would say you can use the second definition if the rest is the same, right?

If you would use `#[serde(untagged)]` with the first struct, would that be the same effect than `Either`?

Reply:
I tried it, but it also doesn't work. The documentation of `Either` says:

> /// It is important to note that this extractor, by necessity, buffers the entire request payload
/// as part of its implementation. Though, it does respect any `PayloadConfig` maximum size limits.

so I guess one cannot combine an `Either` extractor with other extractors. If I require a regular session, everything works with my new struct with the enum inside, but when I add the `Either` to match on both session types, it breaks. In the other handler that use `Either` we do not use an extractor from the payload, but only from the path and query.

Reply:
Maybe you can write another extractor that retrieves a UUID.
Then, it checks whether this UUID is the admin UUID and returns the admin thing-e, or otherwise a normal Session to check.

Reply:
68ba46b28acc48bfeb8fad8ef79447178e18c62b